### PR TITLE
Add self-contained μP support to backbones

### PR DESCRIPTION
**Pull request by joaquin-iturriza**
Introduce a `parametrization="mup"` flag on the Transformer, TransformerV2,
MLP and GraphNet backbones that enables Maximal Update Parametrization (μP)
without any user-managed base/delta models or `.bsh` files.

The new `lloca.mup` subpackage provides:
- `@mup_parametrized`: a class decorator that, when `parametrization="mup"`,
  computes the backbone's μP base shapes during `__init__` by re-instantiating
  it at base/delta widths in memory (deterministic from the constructor args,
  so checkpoint reloads need no extra bookkeeping).
- `make_readout` / μP init helpers (`normal_fanin_`, `reinitialize_mup`).
- `finalize(model)`: marks parameters living outside μP backbones (e.g. a
  frames network) as standard parametrization so a single μP optimizer can
  train the whole model.
- `MuAdam` / `MuAdamW` / `MuSGD` re-exports and a `scale_reps` helper for
  reps-string width axes (GraphNet).

`LLoCaAttention` gains a `parametrization` argument; under μP it scales the
attention logits by 1/d instead of 1/sqrt(d) (via backend-agnostic query
pre-scaling). The standard-parametrization path is unchanged.

`mup` is an optional dependency (`pip install lloca[mup]`). Adds a coordinate-
check test (`tests/lloca/mup`) and a μP guide (`docs/source/mup.rst`).

### CODE REVIEW

File: docs/source/mup.rst [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+       out_channels=out_channels,
+       num_blocks=num_blocks,
+       num_heads=num_heads,        # <- scale this freely; HPs transfer
+       parametrization="mup",
+   )
+
+   # 2. If the trained model has parameters *outside* a μP backbone (e.g. a frames
+   #    network or an input encoder, whose width is fixed), wrap everything and call
+   #    finalize() once to mark those as standard parametrization.
+   model = MyModel(net)            # net is a submodule somewhere inside
```
Comment (spinjo) [inline docs/source/mup.rst:44]: This `MyModel` is not defined here, I think that makes the quick-start example a bit confusing. But I think this aspect of wrapping around the mup core is not needed for a toy example, so maybe you can just have `model=net` instead?

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,18 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+
+### Added
+
+- Self-contained Maximal Update Parametrization (μP) support via the new `lloca.mup`
```
Comment (spinjo) [inline CHANGELOG.md:12]: This should be shortened to follow the style in this file, e.g. 
```
- `parametrization="mup"` option for `Transformer`, `TransformerV2`, `MLP`, `GrpahNet`
```

File: lloca/mup/__init__.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+"""Self-contained Maximal-Update Parametrization (μP) support for LLoCa.
+
+Enable μP on any supported backbone by passing ``parametrization="mup"``; the
```
Comment (spinjo) [inline lloca/mup/__init__.py:3]: This comment block is duplicate with the `mup.rst` docs file, so should be removed

File: lloca/mup/parametrization.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
+"""Self-contained Maximal-Update Parametrization (μP) support for LLoCa backbones.
+
+μP (Yang & Hu, 2022) makes the optimal hyperparameters of a network (most
```
Comment (spinjo) [inline lloca/mup/parametrization.py:3]: This has much overlap with `mup.rst` in the docs, and I think it contains some extra information beyond the docs. I'd propose to include this in the docs, and remove the comment block here. Having it once should be enough.

File: lloca/backbone/attention.py [hunk 5]
```diff
@@ -22,15 +22,29 @@ def __init__(self, attn_reps, num_heads):
             Tensor representation of a single attention head.
         num_heads : int
             Number of attention heads
+        parametrization : str
+            Either ``"sp"`` (standard) or ``"mup"``. Under μP the attention logits
```
Comment (spinjo) [inline lloca/backbone/attention.py:26]: I'd prefer to not use special characters (the mu in muP) in code comments. That makes it inconvenient for non-LLM/non-greek people to edit code. This affects many locations, but I think this can be done quickly with a bash command - or with your favourite LLM.

File: docs/source/mup.rst [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+How it works
+------------
+
+The key observation is that, for these backbones, the μP *base shapes* are a
+deterministic function of the constructor arguments -- the width is just another
+argument (``num_heads`` for the transformers, ``hidden_channels`` for the MLP, the
+``hidden_reps`` multiplicities for ``GraphNet``). So when ``parametrization="mup"`` the
+backbone re-instantiates itself at a small "base" and "delta" width *in memory* and
+computes its base shapes during construction. There are therefore no ``.bsh`` files and
```
Comment (spinjo) [inline docs/source/mup.rst:62]: Any reader who does not know about the original mup repo would be confused about the comment about `.bsh` files and base/delta models. Also you don't mention that you're following the original mup repo approach, but instead of the modular bsh files and base/delta models you hard-code this for a set of manually tested architectures.

I think it would be good to add a paragraph somewhere here that mentions the original mup repo, what their approach is, and how your approach differs and why its better. In particular, it would be good to mention what the bsh files and the base/delta models do so one can appreciate your approach :)

File: tests/lloca/mup/test_mup.py
Comment (spinjo) [inline tests/lloca/mup/test_mup.py:1]: How quickly does this run? I'm a bit scared that the coord check might run a while, especially on the bad CPUs that github uses for their free workflows (this code runs automatically at least on every new `lloca` version release). It should run in less than ~10s, ideally ~1s, if we keep it here.

Maybe a better fit for this is in an `examples/demo_mup.ipynb` notebook. The jupyter notebook format also allows you to display some cached results, so the user directly sees what they should get.

File: lloca/backbone/transformer_v2.py
Comment (spinjo) [inline lloca/backbone/transformer_v2.py:1]: I'm afraid of the day when we make some changes to the `transformer_v2.py` and out of a sudden we get error messages from the `mup` option, even though its disabled by default. Also, I see a risk of people who want to get familiar with LLoCa first and don't need mup yet to get confused when they read these files and wonder if mup is strictly required in LLoCa or so.

To avoid this, I would prefer to add entirely new files `transformer_v2_mup.py` (also for the other classes that you created), and mention early in these files that these are minimal copies that include the mup option. You could then also change the default to `parametrized=mup` in these files. This doesn't hurt the code spirit, because the `lloca/backbone/` folder is meant to be a collection of possible architectures that can be LLoCa-ified.

If you feel strongly about keeping the mup option in the standard files, then we can also try to make that work. But that would require thinking more about how we can make it 100% transparent in the git-diffs of this PR that the mup option is an optional add-on and not used in the default code path.

File: docs/source/mup.rst
Comment (spinjo) [inline docs/source/mup.rst:1]: An important aspect that should be added is a discussion of which parameters can be mup-scaled. The obvious one is the per-head size in the transformers; but for LLoCa specifically scaling the `equivectors` width is also an obvious question. And the story about not scaling with the number of layers. No need to make much more studies to make all of this work, but I think a few honest words on what worked in your tests and what not would be good. This should be a paragraph, or maybe a section, somewhere where it fits.

File: docs/source/mup.rst [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+       mup_delta_shapes={"num_heads": 8},
+   )
+
+The only requirement is that base and delta differ along the width axis; the actual
+model width is independent of both.
+
+Validating a setup (coordinate check)
+-------------------------------------
+
+The standard way to confirm μP is wired correctly is a *coordinate check*: train at a
```
Comment (spinjo) [inline docs/source/mup.rst:98]: These `rst` files support embedded pngs/pdfs. It would be good if you can include some of your coordinate check plots here in a visually appealing way.

File: docs/source/mup.rst [hunk 8]
```diff
@@ -0,0 +1,108 @@
+Maximal Update Parametrization (μP)
+===================================
+
+`Maximal Update Parametrization <https://arxiv.org/abs/2203.03466>`_ (μP) rescales a
```
Comment (spinjo) [inline docs/source/mup.rst:4]: It would be good if you can also cite their repo here, for practitioners thats more relevant than the paper I think.

**Review by spinjo — CHANGES_REQUESTED** [review]
Thanks for the great PR, and sorry for taking so long to review it!

I added a bunch of comments, the main points are to refine the docs and to have separate `transformer_mup.py` etc files for the mup-enabled backbones. We can discuss the points directly on the github PR pages - don't hesitate to ask, there is a good chance that I misunderstood some things about the code.

Also, I'm working on a larger refactor of the code in parallel, that should lead to a `v2.0.0` release at some point in late July / August. But luckily I think there are no conflicts between the two.
